sql/analyzer/optimization_rules.go

```diff
@@ -133,28 +133,19 @@ func moveJoinConditionsToFilter(ctx *sql.Context, a *Analyzer, n sql.Node, scope
 		return node, transform.SameTree, nil
 	}
 
-	// Add a new filter node with all removed predicates above the top level InnerJoin. Or, if there is a filter node
-	// above that, combine into a new filter.
-	selector := func(c transform.Context) bool {
-		switch c.Parent.(type) {
-		case *plan.Filter:
-			return false
+	return transform.NodeWithCtx(node, func(transform.Context) bool { return true }, func(c transform.Context) (sql.Node, transform.TreeIdentity, error) {
+		if c.Node != topJoin {
```

Comment:
Does this function run recursively up the tree, so each join node gets a turn to be the top node?

I'm curious how this interacts with pushdown, could use some more plan tests.

Reply:
There will always be one `topJoin` in a single scope. You would need a subquery, union, etc to get a second scope. If we know the joins are all compact in the tree, we could probably have this whole rule just be a single transform. When we hit a node that isn't a join and the `topJoin` is not nil we combine the conditions turned where filters. The relationship to pushdown is what you were telling me yesterday: this rule only moves filters from join to where clause, pushdown decides how low filters can be pushed.